pyk/src/pyk/proof/proof.py

```diff
@@ -303,13 +307,105 @@ def lines(self) -> list[str]:
         return [line for lines in (summary.lines for summary in self.summaries) for line in lines]
 
 
-class ProofStep: ...
+class FailureInfo: ...
+
+
+class ProverPool(ContextManager['ProverPool'], Generic[P, PS, SR]):
+    _create_prover: Callable[[], Prover[P, PS, SR]]
+    _provers: dict[str, Prover[P, PS, SR]]
+    _executor: Executor
+    _closed: bool
+
+    def __init__(
+        self,
+        create_prover: Callable[[], Prover[P, PS, SR]],
+        *,
+        max_workers: int | None = None,
+    ) -> None:
+        self._create_prover = create_prover
+        self._provers = {}
+        self._executor = ThreadPoolExecutor(max_workers)
+        self._closed = False
 
+    def __enter__(self) -> ProverPool[P, PS, SR]:
+        return self
 
-class StepResult: ...
+    def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
+        self.close()
 
+    def close(self) -> None:
+        self._executor.shutdown()
+        self._closed = True
 
-class FailureInfo: ...
+    def submit(self, proof_step: PS) -> Future[Iterable[SR]]:
+        if self._closed:
+            raise ValueError('ProverPool has been closed')
+        return self._executor.submit(self._with_prover(proof_step))
+
+    def _with_prover(self, proof_step: PS) -> Callable[[], Iterable[SR]]:
+
+        def step() -> Iterable[SR]:
+            thread_name = current_thread().name
+            prover: Prover[P, PS, SR] | None = self._provers.get(thread_name)
+            if prover is None:
+                prover = self._create_prover()
+                self._provers[thread_name] = prover
+            return prover.step_proof(proof_step)
+
+        return step
+
+
+class ParallelProver(Generic[P, PS, SR]):
+
+    def parallel_advance_proof(
```

Comment:
As the class has no state, consider pulling this out as a top-level function.

Reply:
Fixed.